# Limpieza y unión de datos

## Instalación de librerías

In [119]:
import pandas as pd
import geopandas as gpd
from unidecode import unidecode

## Carga de datos

In [120]:
df1 = pd.read_csv(r'C:\Users\EMMY\Desktop\Proyecto\DB Original\Evaluaciones_Agropecuarias_Municipales_EVA_2006-2018.csv')
df2 = pd.read_csv(r'C:\Users\EMMY\Desktop\Proyecto\DB Original\Evaluaciones_Agropecuarias_Municipales___EVA._2019_-_2023._Base_Agr_cola_20240930 copy.csv')

### Primeros datos

In [121]:
df1.head(2)

CÓD. DEP.  DEPARTAMENTO  CÓD. MUN. MUNICIPIO GRUPO DE CULTIVO  \
0         15        BOYACA      15114  BUSBANZA       HORTALIZAS   
1         25  CUNDINAMARCA      25754    SOACHA       HORTALIZAS   

  SUBGRUPO DE CULTIVO CULTIVO DESAGREGACIÓN REGIONAL Y/O SISTEMA PRODUCTIVO  \
0              ACELGA  ACELGA                                        ACELGA   
1              ACELGA  ACELGA                                        ACELGA   

    AÑO PERIODO  Área Sembrada (ha)  Área Cosechada(ha)  Producción(t)  \
0  2006   2006B                   2                   1              1   
1  2006   2006B                  82                  80           1440   

   Rendimiento(t/ha) ESTADO FISICO PRODUCCION NOMBRE CIENTIFICO  \
0                1.0             FRUTO FRESCO     BETA VULGARIS   
1               18.0             FRUTO FRESCO     BETA VULGARIS   

  CICLO DE CULTIVO  
0      TRANSITORIO  
1      TRANSITORIO

In [122]:
df2.head(2)

Código Dane departamento Departamento  Código Dane municipio Municipio  \
0                        13      Bolívar                  13006      Achí   
1                        13      Bolívar                  13006      Achí   

  Grupo cultivo        Subgrupo   Cultivo      Desagregación cultivo   Año  \
0      Frutales  Demás frutales  Aguacate  Aguacate demás variedades  2019   
1      Frutales  Demás frutales  Aguacate  Aguacate demás variedades  2020   

  Periodo  Área sembrada  Área cosechada  Producción  Rendimiento  \
0    2019           80.0            80.0       400.0          5.0   
1    2020           80.0            80.0       400.0          5.0   

  Ciclo del cultivo Estado físico del cultivo  Código del cultivo  \
0        Permanente                 En fresco             2040299   
1        Permanente                 En fresco             2040299   

  Nombre científico del cultivo  
0              Persea americana  
1              Persea americana

In [123]:
df1.columns

Index(['CÓD. DEP.', 'DEPARTAMENTO', 'CÓD. MUN.', 'MUNICIPIO',
       'GRUPO DE CULTIVO', 'SUBGRUPO DE CULTIVO', 'CULTIVO',
       'DESAGREGACIÓN REGIONAL Y/O SISTEMA PRODUCTIVO', 'AÑO', 'PERIODO',
       'Área Sembrada (ha)', 'Área Cosechada(ha)', 'Producción(t)',
       'Rendimiento(t/ha)', 'ESTADO FISICO PRODUCCION', 'NOMBRE CIENTIFICO',
       'CICLO DE CULTIVO'],
      dtype='object')

In [124]:
# Eliminar columnas 'ESTADO FISICO PRODUCCION', 'NOMBRE CIENTIFICO'
df1 = df1.drop(columns=['ESTADO FISICO PRODUCCION', 'NOMBRE CIENTIFICO'])

In [125]:
df2.columns

Index(['Código Dane departamento', 'Departamento', 'Código Dane municipio',
       'Municipio', 'Grupo cultivo', 'Subgrupo', 'Cultivo',
       'Desagregación cultivo', 'Año', 'Periodo', 'Área sembrada',
       'Área cosechada', 'Producción', 'Rendimiento', 'Ciclo del cultivo',
       'Estado físico del cultivo', 'Código del cultivo',
       'Nombre científico del cultivo'],
      dtype='object')

In [126]:
# Eliminar columnas 'Estado físico del cultivo', 'Código del cultivo', 'Nombre científico del cultivo'
df2 = df2.drop(columns=['Estado físico del cultivo', 'Código del cultivo', 'Nombre científico del cultivo'])

In [127]:
df1.columns

Index(['CÓD. DEP.', 'DEPARTAMENTO', 'CÓD. MUN.', 'MUNICIPIO',
       'GRUPO DE CULTIVO', 'SUBGRUPO DE CULTIVO', 'CULTIVO',
       'DESAGREGACIÓN REGIONAL Y/O SISTEMA PRODUCTIVO', 'AÑO', 'PERIODO',
       'Área Sembrada (ha)', 'Área Cosechada(ha)', 'Producción(t)',
       'Rendimiento(t/ha)', 'CICLO DE CULTIVO'],
      dtype='object')

In [128]:
df2.columns

Index(['Código Dane departamento', 'Departamento', 'Código Dane municipio',
       'Municipio', 'Grupo cultivo', 'Subgrupo', 'Cultivo',
       'Desagregación cultivo', 'Año', 'Periodo', 'Área sembrada',
       'Área cosechada', 'Producción', 'Rendimiento', 'Ciclo del cultivo'],
      dtype='object')

## Renombrar las columnas

In [129]:
# Renombrar columnas de df1 
df1.columns = ['Codigo Dane departamento', 'Departamento', 'Codigo Dane municipio',
       'Municipio', 'Grupo cultivo', 'Subgrupo', 'Cultivo',
       'Desagregacion cultivo', 'Año', 'Periodo', 'Area Sembrada (ha)', 'Area Cosechada(ha)', 'Produccion(t)',
       'Rendimiento(t/ha)', 'Ciclo del cultivo']
df1.columns

Index(['Codigo Dane departamento', 'Departamento', 'Codigo Dane municipio',
       'Municipio', 'Grupo cultivo', 'Subgrupo', 'Cultivo',
       'Desagregacion cultivo', 'Año', 'Periodo', 'Area Sembrada (ha)',
       'Area Cosechada(ha)', 'Produccion(t)', 'Rendimiento(t/ha)',
       'Ciclo del cultivo'],
      dtype='object')

In [130]:
# Renombrar columnas de df2
df2.columns = ['Codigo Dane departamento', 'Departamento', 'Codigo Dane municipio',
       'Municipio', 'Grupo cultivo', 'Subgrupo', 'Cultivo',
       'Desagregacion cultivo', 'Año', 'Periodo', 'Area Sembrada (ha)', 'Area Cosechada(ha)', 'Produccion(t)',
       'Rendimiento(t/ha)', 'Ciclo del cultivo']
df2.columns

Index(['Codigo Dane departamento', 'Departamento', 'Codigo Dane municipio',
       'Municipio', 'Grupo cultivo', 'Subgrupo', 'Cultivo',
       'Desagregacion cultivo', 'Año', 'Periodo', 'Area Sembrada (ha)',
       'Area Cosechada(ha)', 'Produccion(t)', 'Rendimiento(t/ha)',
       'Ciclo del cultivo'],
      dtype='object')

In [131]:
df1.columns = df1.columns.str.upper()
df1.columns

Index(['CODIGO DANE DEPARTAMENTO', 'DEPARTAMENTO', 'CODIGO DANE MUNICIPIO',
       'MUNICIPIO', 'GRUPO CULTIVO', 'SUBGRUPO', 'CULTIVO',
       'DESAGREGACION CULTIVO', 'AÑO', 'PERIODO', 'AREA SEMBRADA (HA)',
       'AREA COSECHADA(HA)', 'PRODUCCION(T)', 'RENDIMIENTO(T/HA)',
       'CICLO DEL CULTIVO'],
      dtype='object')

In [132]:
df2.columns = df2.columns.str.upper()
df2.columns

Index(['CODIGO DANE DEPARTAMENTO', 'DEPARTAMENTO', 'CODIGO DANE MUNICIPIO',
       'MUNICIPIO', 'GRUPO CULTIVO', 'SUBGRUPO', 'CULTIVO',
       'DESAGREGACION CULTIVO', 'AÑO', 'PERIODO', 'AREA SEMBRADA (HA)',
       'AREA COSECHADA(HA)', 'PRODUCCION(T)', 'RENDIMIENTO(T/HA)',
       'CICLO DEL CULTIVO'],
      dtype='object')

## Union de los DF

In [133]:
df = pd.concat([df1, df2], axis=0, ignore_index=True)
df.shape[0]

212233

### Seleccionar los datos necesarios

In [134]:
# Seleccionar los datos de córdoba (23) y bolívar (13)
df = df[(df['CODIGO DANE DEPARTAMENTO'] == 23) | (df['CODIGO DANE DEPARTAMENTO'] == 13)]
df['CODIGO DANE DEPARTAMENTO'].unique()

array([13, 23])

In [135]:
df = df.sort_values(['DEPARTAMENTO', 'MUNICIPIO']).reset_index(drop=True)

In [136]:
# Revisar los departamentos seleccionados
df['DEPARTAMENTO'].unique()

array(['BOLIVAR', 'Bolívar', 'CORDOBA', 'Córdoba'], dtype=object)

In [137]:
# Aplicar mapeo específico para corregir registros
df['DEPARTAMENTO'] = df['DEPARTAMENTO'].map({'Bolívar': 'BOLIVAR', 'Córdoba': 'CORDOBA'}).fillna(df['DEPARTAMENTO'])

# Verificar el resultado
df['DEPARTAMENTO'].unique()

array(['BOLIVAR', 'CORDOBA'], dtype=object)

In [138]:
# Aplicar mapeo específico para corregir tildes
df['DEPARTAMENTO'] = df['DEPARTAMENTO'].map({'BOLIVAR': 'BOLÍVAR', 'CORDOBA': 'CÓRDOBA'}).fillna(df['DEPARTAMENTO'])

# Verificar el resultado
df['DEPARTAMENTO'].unique()

array(['BOLÍVAR', 'CÓRDOBA'], dtype=object)

### Poner en mayúscula todo el DF

In [139]:
df = df.apply(lambda col: col.map(str.upper) if col.dtypes == 'object' else col)
df.head()

CODIGO DANE DEPARTAMENTO DEPARTAMENTO  CODIGO DANE MUNICIPIO MUNICIPIO  \
0                        13      BOLÍVAR                  13006      ACHI   
1                        13      BOLÍVAR                  13006      ACHI   
2                        13      BOLÍVAR                  13006      ACHI   
3                        13      BOLÍVAR                  13006      ACHI   
4                        13      BOLÍVAR                  13006      ACHI   

  GRUPO CULTIVO  SUBGRUPO   CULTIVO DESAGREGACION CULTIVO   AÑO PERIODO  \
0      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2007    2007   
1      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2008    2008   
2      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2009    2009   
3      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2010    2010   
4      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2011    2011   

   AREA SEMBRADA (HA)  AREA COSECHADA(HA)  PRODUCCION(T)  RENDIMIENTO(T/HA)  \
0                80.0                80.0          800.0               10.0   
1                80.0                80.0          720.0                9.0   
2                80.0                80.0          800.0               10.0   
3                80.0                80.0          480.0                6.0   
4                80.0                80.0          560.0                7.0   

  CICLO DEL CULTIVO  
0        PERMANENTE  
1        PERMANENTE  
2        PERMANENTE  
3        PERMANENTE  
4        PERMANENTE

In [140]:
# Valores únicos para la columna 'MUNICIPIO'
df['MUNICIPIO'].nunique()

101

### Corregir tíldes

In [141]:
# Diccionario de municipios sin tildes y sus versiones corregidas
correcciones = {
    'ACHI': 'ACHÍ',
    'CORDOBA': 'CÓRDOBA',
    'EL CARMEN DE BOLIVAR': 'EL CARMEN DE BOLÍVAR',
    'EL PEÑON': 'EL PEÑÓN',
    'MAGANGUE': 'MAGANGUÉ',
    'MARIA LA BAJA': 'MARÍA LA BAJA',
    'NOROSI': 'NOROSÍ',
    'RIO VIEJO': 'RÍO VIEJO',
    'SAN CRISTOBAL': 'SAN CRISTÓBAL',
    'SAN MARTIN DE LOBA': 'SAN MARTÍN DE LOBA',
    'SANTA CRUZ DE MOMPOX': 'SANTA CRUZ DE MOMPOX',
    'SIMITI': 'SIMITÍ',
    'TURBANA': 'TURBANÁ',
    'CERETE': 'CERETÉ',
    'CHIMA': 'CHIMÁ',
    'CHINU': 'CHINÚ',
    'CIENAGA DE ORO': 'CIÉNAGA DE ORO',
    'LOS CORDOBAS': 'LOS CÓRDOBAS',
    'MONTELIBANO': 'MONTELÍBANO',
    'MONTERIA': 'MONTERÍA',
    'PURISIMA': 'PURÍSIMA DE LA CONCEPCIÓN',
    'SAHAGUN': 'SAHAGÚN',
    'SAN ANDRES DE SOTAVENTO': 'SAN ANDRÉS DE SOTAVENTO',
    'SAN JOSE DE URE': 'SAN JOSÉ DE URÉ',
    'TUCHIN': 'TUCHÍN'
}

# Reemplazar los valores en la columna 'MUNICIPIO'
df['MUNICIPIO'] = df['MUNICIPIO'].replace(correcciones)

# Verificar las correcciones
df['MUNICIPIO'].nunique()


77

In [142]:
# Corrección manual de inconsistencias específicas
correcciones = {
    'MAIZ': 'MAÍZ',
    'PLATANO': 'PLÁTANO',
    'LIMON': 'LIMÓN',
    'PIMENTON': 'PIMENTÓN',
    'GUANABANA': 'GUANÁBANA',
    'AJONJOLI': 'AJONJOLÍ',
    'MELON': 'MELÓN',
    'AJI': 'AJÍ',
    'NISPERO': 'NÍSPERO',
    'ALGODON': 'ALGODÓN',
    'CAFE': 'CAFÉ',
    'MARAÑON': 'MARAÑÓN',
    'MARACUYA': 'MARACUYÁ',
    'OTROS CITRICOS': 'OTROS CÍTRICOS'
}

# Aplicar las correcciones
df['CULTIVO'] = df['CULTIVO'].replace(correcciones)

## Revisar duplicados

In [143]:
# Revisar el tamaño del DF
df.shape[0]

16354

In [144]:
# Eliminar duplicados
df.drop_duplicates()
df.shape[0]

16354

No hay duplicados

## Características en total del data set

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16354 entries, 0 to 16353
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CODIGO DANE DEPARTAMENTO  16354 non-null  int64  
 1   DEPARTAMENTO              16354 non-null  object 
 2   CODIGO DANE MUNICIPIO     16354 non-null  int64  
 3   MUNICIPIO                 16354 non-null  object 
 4   GRUPO CULTIVO             16354 non-null  object 
 5   SUBGRUPO                  16354 non-null  object 
 6   CULTIVO                   16354 non-null  object 
 7   DESAGREGACION CULTIVO     16354 non-null  object 
 8   AÑO                       16354 non-null  int64  
 9   PERIODO                   16354 non-null  object 
 10  AREA SEMBRADA (HA)        16354 non-null  float64
 11  AREA COSECHADA(HA)        16354 non-null  float64
 12  PRODUCCION(T)             16354 non-null  float64
 13  RENDIMIENTO(T/HA)         16092 non-null  float64
 14  CICLO 

In [146]:
# Cambiar el tipo de datos de ['CODIGO DANE DEPARTAMENTO', 'CODIGO DANE MUNICIPIO', 'AÑO']
df = df.astype({'CODIGO DANE DEPARTAMENTO':'object', 'CODIGO DANE MUNICIPIO':'object', 'AÑO':'object'})
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16354 entries, 0 to 16353
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CODIGO DANE DEPARTAMENTO  16354 non-null  object 
 1   DEPARTAMENTO              16354 non-null  object 
 2   CODIGO DANE MUNICIPIO     16354 non-null  object 
 3   MUNICIPIO                 16354 non-null  object 
 4   GRUPO CULTIVO             16354 non-null  object 
 5   SUBGRUPO                  16354 non-null  object 
 6   CULTIVO                   16354 non-null  object 
 7   DESAGREGACION CULTIVO     16354 non-null  object 
 8   AÑO                       16354 non-null  object 
 9   PERIODO                   16354 non-null  object 
 10  AREA SEMBRADA (HA)        16354 non-null  float64
 11  AREA COSECHADA(HA)        16354 non-null  float64
 12  PRODUCCION(T)             16354 non-null  float64
 13  RENDIMIENTO(T/HA)         16092 non-null  float64
 14  CICLO 

### Estadística descriptiva

In [147]:
df.describe()

AREA SEMBRADA (HA)  AREA COSECHADA(HA)  PRODUCCION(T)  \
count        16354.000000        16354.000000   16354.000000   
mean           457.268609          400.398965    2206.952635   
std            937.883988          854.824803    6395.996295   
min              0.000000            0.000000       0.000000   
25%             25.000000           19.000000      76.000000   
50%            125.000000          100.000000     400.000000   
75%            460.000000          400.000000    1600.000000   
max          11950.000000        13321.000000  156260.000000   

       RENDIMIENTO(T/HA)  
count       16092.000000  
mean            6.485708  
std             7.472140  
min             0.000000  
25%             1.607500  
50%             4.100000  
75%             9.482500  
max           100.470000

Se observan mínimos de 0 para todas las variables y máximos bastante alejados del percentil 75. Esto indica una gran dispersion de los datos

### Valores faltantes

In [148]:
df.isna().sum()

CODIGO DANE DEPARTAMENTO      0
DEPARTAMENTO                  0
CODIGO DANE MUNICIPIO         0
MUNICIPIO                     0
GRUPO CULTIVO                 0
SUBGRUPO                      0
CULTIVO                       0
DESAGREGACION CULTIVO         0
AÑO                           0
PERIODO                       0
AREA SEMBRADA (HA)            0
AREA COSECHADA(HA)            0
PRODUCCION(T)                 0
RENDIMIENTO(T/HA)           262
CICLO DEL CULTIVO             0
dtype: int64

Existen 262 registros en los que RENDIMIENTO(T/HA) no tine información

In [149]:
# Ver los registros con valores NaN 
df_nan = df[df.isna().any(axis=1)]
df_nan

CODIGO DANE DEPARTAMENTO DEPARTAMENTO CODIGO DANE MUNICIPIO  \
227                         13      BOLÍVAR                 13030   
228                         13      BOLÍVAR                 13030   
237                         13      BOLÍVAR                 13030   
238                         13      BOLÍVAR                 13030   
339                         13      BOLÍVAR                 13042   
...                        ...          ...                   ...   
13283                       23      CÓRDOBA                 23815   
13385                       23      CÓRDOBA                 23855   
13386                       23      CÓRDOBA                 23855   
13387                       23      CÓRDOBA                 23855   
13388                       23      CÓRDOBA                 23855   

               MUNICIPIO          GRUPO CULTIVO         SUBGRUPO  \
227    ALTOS DEL ROSARIO      OTROS PERMANENTES            CACAO   
228    ALTOS DEL ROSARIO      OTROS PERMANENTES            CACAO   
237    ALTOS DEL ROSARIO      OTROS PERMANENTES            CACAO   
238    ALTOS DEL ROSARIO      OTROS PERMANENTES            CACAO   
339               ARENAL            OLEAGINOSAS  PALMA DE ACEITE   
...                  ...                    ...              ...   
13283             TUCHÍN  TUBERCULOS Y PLATANOS           BATATA   
13385           VALENCIA             FORESTALES           CAUCHO   
13386           VALENCIA             FORESTALES           CAUCHO   
13387           VALENCIA             FORESTALES           CAUCHO   
13388           VALENCIA             FORESTALES           CAUCHO   

               CULTIVO DESAGREGACION CULTIVO   AÑO PERIODO  \
227              CACAO                 CACAO  2007    2007   
228              CACAO                 CACAO  2008    2008   
237              CACAO                 CACAO  2017    2017   
238              CACAO                 CACAO  2018    2018   
339    PALMA DE ACEITE       PALMA DE ACEITE  2007    2007   
...                ...                   ...   ...     ...   
13283           BATATA                BATATA  2015   2015A   
13385           CAUCHO                CAUCHO  2013    2013   
13386           CAUCHO                CAUCHO  2014    2014   
13387           CAUCHO                CAUCHO  2015    2015   
13388           CAUCHO                CAUCHO  2016    2016   

       AREA SEMBRADA (HA)  AREA COSECHADA(HA)  PRODUCCION(T)  \
227                  20.0                 0.0            0.0   
228                  32.0                 0.0            0.0   
237                  20.0                 0.0            0.0   
238                  20.0                 0.0            0.0   
339                  85.0                 0.0            0.0   
...                   ...                 ...            ...   
13283                27.0                 0.0            0.0   
13385                80.0                 0.0            0.0   
13386               100.0                 0.0            0.0   
13387               100.0                 0.0            0.0   
13388                35.0                 0.0            0.0   

       RENDIMIENTO(T/HA) CICLO DEL CULTIVO  
227                  NaN        PERMANENTE  
228                  NaN        PERMANENTE  
237                  NaN        PERMANENTE  
238                  NaN        PERMANENTE  
339                  NaN        PERMANENTE  
...                  ...               ...  
13283                NaN       TRANSITORIO  
13385                NaN        PERMANENTE  
13386                NaN        PERMANENTE  
13387                NaN        PERMANENTE  
13388                NaN        PERMANENTE  

[262 rows x 15 columns]

In [150]:
df_nan['AREA COSECHADA(HA)'].unique()

array([0.])

In [151]:
df_nan['PRODUCCION(T)'].unique()

array([0.])

Se puede observar que tanto AREA COSECHADA(HA) como PRODUCCION(T) son 0 para los registros NaN de RENDIMIENTO(T/HA). Para un mejor manejo de los datos se van a reemplazar los NaN con 0

In [152]:
# Remplazar NaN con 0
df['RENDIMIENTO(T/HA)'] = df['RENDIMIENTO(T/HA)'].fillna(0)

# Revisar NaN
df.isna().sum()

CODIGO DANE DEPARTAMENTO    0
DEPARTAMENTO                0
CODIGO DANE MUNICIPIO       0
MUNICIPIO                   0
GRUPO CULTIVO               0
SUBGRUPO                    0
CULTIVO                     0
DESAGREGACION CULTIVO       0
AÑO                         0
PERIODO                     0
AREA SEMBRADA (HA)          0
AREA COSECHADA(HA)          0
PRODUCCION(T)               0
RENDIMIENTO(T/HA)           0
CICLO DEL CULTIVO           0
dtype: int64

### Revisar la coherencia de los datos
- AREA SEMBRADA (HA) >=	AREA COSECHADA(HA)
- 

In [153]:
# AREA SEMBRADA (HA) >=	AREA COSECHADA(HA)
df[df['AREA SEMBRADA (HA)'] < df['AREA COSECHADA(HA)']]

CODIGO DANE DEPARTAMENTO DEPARTAMENTO CODIGO DANE MUNICIPIO MUNICIPIO  \
5100                        13      BOLÍVAR                 13006      ACHÍ   
5101                        13      BOLÍVAR                 13006      ACHÍ   
5102                        13      BOLÍVAR                 13006      ACHÍ   
5106                        13      BOLÍVAR                 13006      ACHÍ   
5107                        13      BOLÍVAR                 13006      ACHÍ   
...                        ...          ...                   ...       ...   
16293                       23      CÓRDOBA                 23855  VALENCIA   
16294                       23      CÓRDOBA                 23855  VALENCIA   
16303                       23      CÓRDOBA                 23855  VALENCIA   
16333                       23      CÓRDOBA                 23855  VALENCIA   
16351                       23      CÓRDOBA                 23855  VALENCIA   

             GRUPO CULTIVO             SUBGRUPO          CULTIVO  \
5100              CEREALES             CEREALES            ARROZ   
5101              CEREALES             CEREALES            ARROZ   
5102              CEREALES             CEREALES            ARROZ   
5106              CEREALES             CEREALES            ARROZ   
5107              CEREALES             CEREALES            ARROZ   
...                    ...                  ...              ...   
16293             CEREALES             CEREALES             MAÍZ   
16294             CEREALES             CEREALES             MAÍZ   
16303  RAÍCES Y TUBÉRCULOS  RAÍCES Y TUBÉRCULOS             ÑAME   
16333           HORTALIZAS  HORTALIZAS DE FRUTO  PEPINO COHOMBRO   
16351  RAÍCES Y TUBÉRCULOS  RAÍCES Y TUBÉRCULOS             YUCA   

           DESAGREGACION CULTIVO   AÑO PERIODO  AREA SEMBRADA (HA)  \
5100                 ARROZ RIEGO  2022   2022B               143.0   
5101         ARROZ SECANO MANUAL  2022   2022B               630.0   
5102     ARROZ SECANO MECANIZADO  2022   2022B              6660.0   
5106                 ARROZ RIEGO  2023   2023B               143.0   
5107         ARROZ SECANO MANUAL  2023   2023B               630.0   
...                          ...   ...     ...                 ...   
16293  MAÍZ AMARILLO TECNIFICADO  2023   2023B               300.0   
16294    MAÍZ BLANCO TECNIFICADO  2023   2023B               700.0   
16303                ÑAME ESPINO  2022   2022B                 4.0   
16333            PEPINO COHOMBRO  2022   2022A                 2.0   
16351     YUCA CONSUMO EN FRESCO  2022   2022B                50.0   

       AREA COSECHADA(HA)  PRODUCCION(T)  RENDIMIENTO(T/HA) CICLO DEL CULTIVO  
5100                742.0         2968.0                4.0       TRANSITORIO  
5101               1200.0         3600.0                3.0       TRANSITORIO  
5102              13321.0        46623.5                3.5       TRANSITORIO  
5106                429.0         1716.0                4.0       TRANSITORIO  
5107               1030.0         3090.0                3.0       TRANSITORIO  
...                   ...            ...                ...               ...  
16293               385.0         1732.5                4.5       TRANSITORIO  
16294               795.0         3577.5                4.5       TRANSITORIO  
16303                 7.0           14.0                2.0       TRANSITORIO  
16333                 5.0            7.5                1.5       TRANSITORIO  
16351                70.0          840.0               12.0       TRANSITORIO  

[627 rows x 15 columns]

Existen 627 registros en los que AREA SEMBRADA (HA) < AREA COSECHADA(HA), lo cual no tiene sentito <br>
Se eliminan estos registros

In [154]:
df = df[df['AREA SEMBRADA (HA)'] >= df['AREA COSECHADA(HA)']]
df.shape[0]

15727

Quedamos con 15727 registros 

## Nueva revisión de estadística descriptiva

In [155]:
df.describe()

AREA SEMBRADA (HA)  AREA COSECHADA(HA)  PRODUCCION(T)  \
count        15727.000000        15727.000000   15727.000000   
mean           466.856797          396.268574    2190.196082   
std            949.812460          852.309949    6438.652888   
min              0.100000            0.000000       0.000000   
25%             25.500000           18.000000      75.000000   
50%            130.000000           98.000000     390.000000   
75%            475.000000          390.000000    1560.000000   
max          11950.000000        11950.000000  156260.000000   

       RENDIMIENTO(T/HA)  
count       15727.000000  
mean            6.412599  
std             7.541657  
min             0.000000  
25%             1.500000  
50%             4.000000  
75%             9.000000  
max           100.470000

In [156]:
# Registros con 'AREA SEMBRADA (HA)' inferior a 1
df[df['AREA SEMBRADA (HA)'] < 1].shape[0]

20

In [157]:
df.columns

Index(['CODIGO DANE DEPARTAMENTO', 'DEPARTAMENTO', 'CODIGO DANE MUNICIPIO',
       'MUNICIPIO', 'GRUPO CULTIVO', 'SUBGRUPO', 'CULTIVO',
       'DESAGREGACION CULTIVO', 'AÑO', 'PERIODO', 'AREA SEMBRADA (HA)',
       'AREA COSECHADA(HA)', 'PRODUCCION(T)', 'RENDIMIENTO(T/HA)',
       'CICLO DEL CULTIVO'],
      dtype='object')

In [158]:
# Poner columnas en minúscula para mayor facilidad
data = df.copy()
data.columns = data.columns.str.lower()

In [159]:
# Renombrar columnas para mayor facilidad
data.columns = ['codigo_dane_departamento', 'departamento', 'codigo_dane_municipio',
       'municipio', 'grupo_cultivo', 'subgrupo', 'cultivo',
       'desagregacion_cultivo', 'año', 'periodo', 'area_sembrada',
       'area_cosechada', 'produccion', 'rendimiento',
       'ciclo_cultivo']

In [160]:
data.columns

Index(['codigo_dane_departamento', 'departamento', 'codigo_dane_municipio',
       'municipio', 'grupo_cultivo', 'subgrupo', 'cultivo',
       'desagregacion_cultivo', 'año', 'periodo', 'area_sembrada',
       'area_cosechada', 'produccion', 'rendimiento', 'ciclo_cultivo'],
      dtype='object')

In [161]:
# Poner el df en formato Título para mejor lectura
data = data.map(lambda x: x.title() if isinstance(x, str) else x)
data.head()

codigo_dane_departamento departamento  codigo_dane_municipio municipio  \
0                        13      Bolívar                  13006      Achí   
1                        13      Bolívar                  13006      Achí   
2                        13      Bolívar                  13006      Achí   
3                        13      Bolívar                  13006      Achí   
4                        13      Bolívar                  13006      Achí   

  grupo_cultivo  subgrupo   cultivo desagregacion_cultivo   año periodo  \
0      Frutales  Aguacate  Aguacate              Aguacate  2007    2007   
1      Frutales  Aguacate  Aguacate              Aguacate  2008    2008   
2      Frutales  Aguacate  Aguacate              Aguacate  2009    2009   
3      Frutales  Aguacate  Aguacate              Aguacate  2010    2010   
4      Frutales  Aguacate  Aguacate              Aguacate  2011    2011   

   area_sembrada  area_cosechada  produccion  rendimiento ciclo_cultivo  
0           80.0            80.0       800.0         10.0    Permanente  
1           80.0            80.0       720.0          9.0    Permanente  
2           80.0            80.0       800.0         10.0    Permanente  
3           80.0            80.0       480.0          6.0    Permanente  
4           80.0            80.0       560.0          7.0    Permanente

## Guardar df en un nuevo archivo

In [175]:
data['cultivo'].unique()

array(['Aguacate', 'Ahuyama', 'Arroz', 'Cacao', 'Caña Panelera', 'Coco',
       'Guayaba', 'Maíz', 'Mango', 'Patilla', 'Plátano', 'Yuca', 'Zapote',
       'Ají', 'Algodón', 'Palma De Aceite', 'Papaya', 'Piña', 'Ñame',
       'Frijol', 'Caucho', 'Berenjena', 'Cilantro', 'Ciruela', 'Col',
       'Guanábana', 'Hortalizas Varias', 'Melón', 'Sorgo',
       'Pepino Cohombro', 'Pimentón', 'Tomate', 'Naranja', 'Habichuela',
       'Limón', 'Ajonjolí', 'Tabaco Negro', 'Tabaco Rubio', 'Citricos',
       'Jatropha', 'Café', 'Higuerilla', 'Eucalipto Baby Blue',
       'Maiz Forrajero', 'Pitahaya', 'Níspero', 'Caña', 'Batata',
       'Otras Hortalizas', 'Millo', 'Sacha Inchi', 'Tabaco', 'Mandarina',
       'Otros Cítricos', 'Pomelo', 'Achiote', 'Maracuyá', 'Calabaza',
       'Marañón', 'Caña Flecha', 'Guandul', 'Mamoncillo',
       'Otros Frutales', 'Soya', 'Calabacín, Calabaza',
       'Plantas Aromáticas', 'Lima',
       'Otros Cultivos Tropicales Tradicionales'], dtype=object)

In [162]:
data.to_csv('datos.csv', index=False)

## Exploración adicional de datos

### Registros con producción > 0

In [163]:
df1 = df[df['PRODUCCION(T)'] == 0]
df1.shape[0]

653

In [164]:
df1[df1['RENDIMIENTO(T/HA)'] != 0]

CODIGO DANE DEPARTAMENTO DEPARTAMENTO CODIGO DANE MUNICIPIO  \
11318                       23      CÓRDOBA                 23570   

          MUNICIPIO      GRUPO CULTIVO SUBGRUPO CULTIVO DESAGREGACION CULTIVO  \
11318  PUEBLO NUEVO  OTROS PERMANENTES    CACAO   CACAO                 CACAO   

        AÑO PERIODO  AREA SEMBRADA (HA)  AREA COSECHADA(HA)  PRODUCCION(T)  \
11318  2009    2009                 1.0                 1.0            0.0   

       RENDIMIENTO(T/HA) CICLO DEL CULTIVO  
11318               0.44        PERMANENTE

### Revisar si la columna rendimiento cumple con la condición t/ha

In [165]:
# Nueva columna con la formula
df1 = df.copy()
df1['REND'] = (df1['PRODUCCION(T)'] / df1['AREA COSECHADA(HA)']).round(2)

In [166]:
# Rellenar valores NaN de la nueva columna con 0
df1['REND']=df1['REND'].fillna(0)

In [167]:
# Nueva columna que indica si 'RENDIMIENTO(T/HA)' cumple con la condición t/ha_cosechadas 
df1['cumple'] = df1.apply(lambda x: 'si' if x['REND'] == x['RENDIMIENTO(T/HA)'] else 'no', axis=1) # Se usa apply porque es por columnas no por elemento
df1.head()

CODIGO DANE DEPARTAMENTO DEPARTAMENTO CODIGO DANE MUNICIPIO MUNICIPIO  \
0                       13      BOLÍVAR                 13006      ACHÍ   
1                       13      BOLÍVAR                 13006      ACHÍ   
2                       13      BOLÍVAR                 13006      ACHÍ   
3                       13      BOLÍVAR                 13006      ACHÍ   
4                       13      BOLÍVAR                 13006      ACHÍ   

  GRUPO CULTIVO  SUBGRUPO   CULTIVO DESAGREGACION CULTIVO   AÑO PERIODO  \
0      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2007    2007   
1      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2008    2008   
2      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2009    2009   
3      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2010    2010   
4      FRUTALES  AGUACATE  AGUACATE              AGUACATE  2011    2011   

   AREA SEMBRADA (HA)  AREA COSECHADA(HA)  PRODUCCION(T)  RENDIMIENTO(T/HA)  \
0                80.0                80.0          800.0               10.0   
1                80.0                80.0          720.0                9.0   
2                80.0                80.0          800.0               10.0   
3                80.0                80.0          480.0                6.0   
4                80.0                80.0          560.0                7.0   

  CICLO DEL CULTIVO  REND cumple  
0        PERMANENTE  10.0     si  
1        PERMANENTE   9.0     si  
2        PERMANENTE  10.0     si  
3        PERMANENTE   6.0     si  
4        PERMANENTE   7.0     si

In [168]:
df1.shape

(15727, 17)

In [169]:
df1[df1['cumple'] == 'no'].shape[0]

870

Se tienen 870 registros que no cumplen con la condición  t/ha_cosechadas

### Revisión profunda de los registros que no cumplen la condición

In [170]:
df2 = df1[df1['cumple'] == 'no'].copy()

In [171]:
# Nueva columna que indica la diferencia entre el rendimiento que está en el df original y el rendimiento calculado con la formula
df2['dif'] = df2['REND'] - df2['RENDIMIENTO(T/HA)']

In [172]:
df2.head()

CODIGO DANE DEPARTAMENTO DEPARTAMENTO CODIGO DANE MUNICIPIO  \
18                        13      BOLÍVAR                 13006   
79                        13      BOLÍVAR                 13006   
111                       13      BOLÍVAR                 13006   
112                       13      BOLÍVAR                 13006   
213                       13      BOLÍVAR                 13030   

             MUNICIPIO      GRUPO CULTIVO SUBGRUPO  CULTIVO  \
18                ACHÍ           CEREALES    ARROZ    ARROZ   
79                ACHÍ  OTROS PERMANENTES    CACAO    CACAO   
111               ACHÍ           FRUTALES  GUAYABA  GUAYABA   
112               ACHÍ           FRUTALES  GUAYABA  GUAYABA   
213  ALTOS DEL ROSARIO           CEREALES    ARROZ    ARROZ   

    DESAGREGACION CULTIVO   AÑO PERIODO  AREA SEMBRADA (HA)  \
18            ARROZ RIEGO  2016   2016B               387.0   
79                  CACAO  2016    2016                35.0   
111               GUAYABA  2013    2013                13.0   
112               GUAYABA  2014    2014                13.0   
213   ARROZ SECANO MANUAL  2014   2014B                60.0   

     AREA COSECHADA(HA)  PRODUCCION(T)  RENDIMIENTO(T/HA) CICLO DEL CULTIVO  \
18                387.0         1503.0               3.89       TRANSITORIO   
79                 35.0           18.0               0.50        PERMANENTE   
111                13.0           72.0               5.50        PERMANENTE   
112                13.0           72.0               5.50        PERMANENTE   
213                55.0           83.0               1.50       TRANSITORIO   

     REND cumple   dif  
18   3.88     no -0.01  
79   0.51     no  0.01  
111  5.54     no  0.04  
112  5.54     no  0.04  
213  1.51     no  0.01

In [173]:
# Número de registros que tienen una diferencia superior a 1
df2[abs(df2['dif'])>1].shape[0]

50

# Extraer el nombre de los archivos de una carpeta 
## zip

```python
import zipfile

# Ruta al archivo ZIP
ruta_zip = r'C:\Users\EMMY\Desktop\Proyecto\DB Original\Subregiones_-_Provincias_de_Colombia.zip'

# Abrir el archivo ZIP y listar los nombres
with zipfile.ZipFile(ruta_zip, 'r') as archivo_zip:
    nombres_archivos = archivo_zip.namelist()

# Imprimir los nombres de los archivos
for nombre in nombres_archivos:
    print(nombre)


## Carpeta normal

In [174]:
import os

# Ruta a la carpeta
ruta_carpeta = r'C:\Users\EMMY\Desktop\Proyecto\DB Original'

# Listar solo archivos (excluyendo carpetas)
archivos = [archivo for archivo in os.listdir(ruta_carpeta) if os.path.isfile(os.path.join(ruta_carpeta, archivo))]

# Imprimir los nombres de los archivos
for archivo in archivos:
    print(archivo)


Evaluaciones_Agropecuarias_Municipales_EVA_2006-2018.csv
Evaluaciones_Agropecuarias_Municipales___EVA._2019_-_2023._Base_Agr_cola_20240930 copy.csv
mapa.cpg
mapa.dbf
mapa.prj
mapa.shp
mapa.shx
mapas.ipynb
subregiones.pdf
Subregiones_-_Provincias_de_Colombia.csv
Subregiones_-_Provincias_de_Colombia.geojson


### Transformar .shp a .json

```python
# Cargar el Shapefile
shapefile = r'C:\Users\EMMY\Desktop\Proyecto\DB Original\Subregiones_-_Provincias_de_Colombia\Subregiones_-_Provincias_de_Colombia.shp'
data = gpd.read_file(shapefile)

# Exportar a GeoJSON
data.to_file("Subregiones_Colombia.json", driver="GeoJSON")